controllers/cloud_config_sync_controller.go

```diff
@@ -0,0 +1,196 @@
+package controllers
+
+import (
+	"context"
+	"fmt"
+	configv1 "github.com/openshift/api/config/v1"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/client-go/tools/record"
+	"k8s.io/klog"
+	"reflect"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/builder"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/event"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+)
+
+const (
+	openshiftConfigNamespace        = "openshift-config"
+	openshiftManagedConfigNamespace = "openshift-config-managed"
+	managedCloudConfigMapName       = "kube-cloud-config"
+
+	cloudConfigMapName = "cloud-conf"
+	defaultConfigKey   = "cloud.conf"
+)
+
+type CloudConfigReconciler struct {
+	client.Client
+	Scheme                 *runtime.Scheme
+	Recorder               record.EventRecorder
+	watcher                ObjectWatcher
+	TargetNamespace        string
+	InfrastructureResource *configv1.Infrastructure
+}
+
+func (r *CloudConfigReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	klog.Infof("%s emitted event, syncing cloud-conf ConfigMap", req)
+	var err error
+
+	sourceCM := &corev1.ConfigMap{}
+	sourceConfigMapKey := client.ObjectKey{Name: req.Name, Namespace: req.Namespace}
+	if req.Namespace == r.TargetNamespace {
+		klog.Infof("target cloud-config somewhat changed, syncing with cloud-config from openshift-config namespace")
+		sourceConfigMapKey = client.ObjectKey{
+			Name: r.InfrastructureResource.Spec.CloudConfig.Name, Namespace: openshiftConfigNamespace,
+		}
+	}
+	err = r.Get(ctx, sourceConfigMapKey, sourceCM)
+	if err != nil {
+		klog.Errorf("unable to get source cloud-config: %v", err)
+		return ctrl.Result{}, err
+	}
+	sourceCM, err = r.prepareSourceConfigMap(sourceCM)
```

Comment:
Might be less error prone/confusing if we use a new variable here instead?
```suggestion
	preparedCM, err := r.prepareSourceConfigMap(sourceCM)
```